lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,817 @@
+  // This allows for the emission of transition arcs for the regs
+  // to be deferred to the end of the pass.
+  // This is necessary, as we need to wait for the `next` operation to
+  // have been converted to btor2 before we can emit the transition.
+  SmallVector<Operation *> regOps;
+
+  // Set of often reused strings in btor2 emission (to avoid typos and enable
+  // auto-complete)
+  static constexpr StringLiteral sortStr = "sort";
+  static constexpr StringLiteral bitvecStr = "bitvec";
+  static constexpr StringLiteral inputStr = "input";
+  static constexpr StringLiteral resetStr = "reset";
+  static constexpr StringLiteral outputStr = "output";
+  static constexpr StringLiteral zeroStr = "zero";
+  static constexpr StringLiteral oneStr = "one";
+  static constexpr StringLiteral constStr = "const";
+  static constexpr StringLiteral constdStr = "constd";
+  static constexpr StringLiteral consthStr = "consth";
+  static constexpr StringLiteral sliceStr = "slice";
+  static constexpr StringLiteral uextStr = "uext";
+  static constexpr StringLiteral addStr = "add";
+  static constexpr StringLiteral subStr = "sub";
+  static constexpr StringLiteral mulStr = "mul";
+  static constexpr StringLiteral andStr = "and";
+  static constexpr StringLiteral orStr = "or";
+  static constexpr StringLiteral xorStr = "xor";
+  static constexpr StringLiteral sllStr = "sll";
+  static constexpr StringLiteral srlStr = "srl"; // a.k.a. unsigned right shift
+  static constexpr StringLiteral sraStr = "sra"; // a.k.a. signed right shift
+  static constexpr StringLiteral sdivStr = "sdiv";
+  static constexpr StringLiteral udivStr = "udiv";
+  static constexpr StringLiteral smodStr = "smod";
+  static constexpr StringLiteral concatStr = "concat";
+  static constexpr StringLiteral notStr = "not";
+  static constexpr StringLiteral neqStr = "neq";
+  static constexpr StringLiteral hwNeqStr = "ne";
+  static constexpr StringLiteral iteStr = "ite";
+  static constexpr StringLiteral impliesStr = "implies"; // logical implication
+  static constexpr StringLiteral stateStr = "state";     // Register state
+  static constexpr StringLiteral nextStr = "next"; // Register state transition
+  static constexpr StringLiteral badStr = "bad";
+  static constexpr StringLiteral constraintStr = "constraint";
+  static constexpr StringLiteral wsStr = " ";  // WhiteSpace
+  static constexpr StringLiteral nlStr = "\n"; // NewLine
+
+  // Constants used during the conversion
+  static constexpr size_t noLID = -1UL;
+  static constexpr int64_t noWidth = -1L;
+
+  /// Field helper functions
+
+  // Checks if an operation was declared
+  // If so, its lid will be returned
+  // Otherwise -1 will be returned
+  size_t getOrCreateOpLID(Operation *op) {
+    if (op == nullptr)
+      return noLID;
+
+    // Look for the original operation declaration
+    // Make sure that wires are considered when looking for an lid
+    Operation *defOp = getOpAlias(op);
+    if (opLIDMap.contains(defOp)) {
```

Comment:
this double, wait quadruple lookup can be removed.
```
auto& f = opLIDMap[defOp];
if (!f) f= lid;
return f;
```
doing an .at after an insertion is gratuitously wasteful.